TDP-3037 Recording feature (#226)
Bandwidth/bandwidth.github.io: voice/bxml/verbs/record.md

{% method %}
## XML: `<Record>`
The Record verb allows a segment of audio to be recorded during a call. At the end of the recording, a [Record Complete](../callbacks/recordComplete.md) event is generated.

### Attributes

| Attribute                    | Description                                                                                                                                                                                                            |
|:-----------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | recordCompleteUrl            | (optional) URL to send the [Record Complete](../callbacks/recordComplete.md) event to and request new BXML.                                                                                                            |

Reviewer: ```suggestion
| recordCompleteUrl            | (optional) URL to send the [Record Complete](../callbacks/recordComplete.md) event to once it has ended. Accepts BXML.                                                                                                            |
```

Reviewer: I'm just now realizing that we use `recordingAvailable` but also `recordComplete`. I'd almost like to change the latter to `recordingComplete` as well, but that's not exactly a small change.
:thinking: 
I suppose it makes sense to keep it as-is, since it's specifically for the `Record` verb and not other recordings (viz., those started by `StartRecording`)

Author: There was a discussion about it in the solution doc. In the end `recordComplete` was chosen.

| recordCompleteMethod         | (optional) The HTTP method to use for the request to `recordCompleteUrl`. GET or POST. Default value is POST.                                                                                                          |

> | recordingAvailableUrl        | (optional) URL to send the [Recording Available](../callbacks/recordingAvailable.md) event to.                                                                                                                         |

Reviewer: ```suggestion
| recordingAvailableUrl        | (optional) URL to send the [Recording Available](../callbacks/recordingAvailable.md) event to once it has been processed. Does not accept BXML.                                                                                                                         |
```

| recordingAvailableMethod     | (optional) The HTTP method to use for the request to `recordingAvailableUrl`. GET or POST. Default value is POST.                                                                                                      |

> | username                     | (optional) The username to send in the HTTP request to `recordCompleteUrl` or `recordingAvailableUrl`. If username and password are specified, then https is required.                                                 |

Reviewer: ```suggestion
| username                     | (optional) The username to send in the HTTP request to `recordCompleteUrl` or `recordingAvailableUrl`. If specified, the `recordingAvailableUrl` must be TLS-encrypted (i.e., `https`).                                                 |
```

Reviewer: Specifying just one also requires `https`, right?

Author: If only one is specified then it is ignored completely :-o
See `HttpUtils#generateBasicAuthentication`.

> | password                     | (optional) The password to send in the HTTP request to `recordCompleteUrl` or `recordingAvailableUrl`. If username and password are specified, then https is required.                                                 |

Reviewer: ```suggestion
| password                     | (optional) The password to send in the HTTP request to `recordCompleteUrl` or `recordingAvailableUrl`. If specified, the `recordingAvailableUrl` must be TLS-encrypted (i.e., `https`).                                                 |
```

| tag                          | (optional) A custom string that will be sent with this and all future callbacks unless overwritten by a future `tag` attribute or cleared.<br><br>May be cleared by setting `tag=""`<br><br>Max length 256 characters. |
| terminatingDigits            | (optional) When pressed, this digit will terminate the recording. Default value is `“#”`.                                                                                                                              |

> | maxDuration                  | (optional) Maximum length of recording (in seconds). Max 10800 (3 hours). Default value is 60.                                                                                                                         |

Reviewer: ```suggestion
| maxDuration                  | (optional) Maximum length of recording (in seconds). Max is 10800 (3 hours). Default value is 60.                                                                                                                         |
```

Reviewer: I wish we took ISO 8601 durations :slightly_frowning_face:

> | fileFormat                   | (optional) The format that the recording will be saved in - `mp3` or `wav`.  Default value is `wav`.                                                                                                                   |

Reviewer: This sentence is hard to express casually and correctly. :thinking: 
```suggestion
| fileFormat                   | (optional) The format that the recording will be saved in: `mp3` or `wav`.  Default value is `wav`.                                                                                                                   |
```

> | fileFormat                   | (optional) The format that the recording will be saved in: `mp3` or `wav`.  Default value is `wav`.                                                                                                                    |

Reviewer: ```suggestion
| fileFormat                   | (optional) The audio format that the recording will be saved as: `mp3` or `wav`.  Default value is `wav`.                                                                                                                    |
```

> The caller can press a `terminatingDigit` to stop the recording.

Reviewer: ```suggestion
The caller can press one of the `terminatingDigits` to stop the recording.
```

Reviewer: :thinking: `terminatingDigit` should probably be `terminatingDigits`

> A `maxDuration` can be specified to stop recording after a specified period.

Reviewer: ```suggestion
A `maxDuration` can be specified to stop recording after a specified period of time.
```

> If the `recordCompleteUrl` attribute is specified, the [Recording Complete](../callbacks/recordComplete.md) event is sent to the `recordCompleteUrl` and the BXML returned by that callback is executed next and all verbs following the `<Record>` tag will be ignored. If no `recordCompleteUrl` is specified, execution of verbs following the `<Record>` tag continues. The recording will still be available on the server.

Reviewer: ```suggestion
If the `recordCompleteUrl` attribute is specified, then the [Recording Complete](../callbacks/recordComplete.md) event is sent to the `recordCompleteUrl` and the BXML returned by that callback is executed next, replacing all verbs following the `<Record>` tag in the original BXML document. If no `recordCompleteUrl` is specified, verb execution continues normally (i.e., after the `<Record>` tag in the original BXML document).
```

> If the `recordingAvailableUrl` attribute is specified, the [Recording Available](../callbacks/recordingAvailable.md) event is sent to the URL once the recording is available for download. No BXML is expected.

Reviewer: ```suggestion
If the `recordingAvailableUrl` attribute is specified, the [Recording Available](../callbacks/recordingAvailable.md) event is sent to the URL once the recording is available for download. BXML returned in response to this callback will be ignored.
```

### Callbacks Received

| Callbacks                                                         | Can reply with more BXML |
|:------------------------------------------------------------------|:-------------------------|
| [Record Complete](../callbacks/recordComplete.md)                 | Yes                      |
| [Recording Available](../callbacks/recordingAvailable.md)         | No                       |

{% common %}
#### Example: Redirect Verb
This shows how to use Bandwidth XML to record a phone call.

```XML
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <SpeakSentence voice="bridget">Please leave your message after the beep</SpeakSentence>
    <PlayAudio>http://audio.url/beep.wav</PlayAudio>

>     <Record recordCompleteUrl="https://myapp.com/nextBXML" maxDuration="10"/>

Reviewer: ```suggestion
    <Record recordCompleteUrl="https://myapp.com/nextBXML" maxDuration="10"/>
    <!-- Any verbs after this point would be replaced by those returned by the recordCompleteUrl -->
```
